### ISSUE TITLE
Error: Signature mismatch:

### ISSUE THREAD

**Issue by nomeata** [QuickChick/QuickChick#87]
I am trying to use QuickChick for the first time, and I stumble about this:
```
File "/tmp/QuickChick/145645/QuickChick30e51e.ml", line 3765, characters 18-26:
Error: Signature mismatch:
       ...
       Values do not match:
         val compare : char list -> char list -> 'a compare0
       is not included in
         val compare : t -> t -> t compare0
       File "/tmp/QuickChick/145645/QuickChick30e51e.ml", line 1848, characters 2-36:
         Expected declaration
       File "/tmp/QuickChick/145645/QuickChick30e51e.ml", line 1949, characters 10-17:
         Actual declaration
```
I have not yet tried to create a minimal example, but maybe you can tell what’s wrong based on this error message already.

**Comment by nomeata** [comment]
The problem seems to be with `compare` in this file:
https://github.com/QuickChick/QuickChick/blob/master/src/StringOT.v#L157

The extracted ml code has

```
type 'x compare0 =
| LT
| EQ
| GT
```
(note the phantom type parameter), and
```
module type OrderedType =
 sig
  type t

  val compare : t -> t -> t compare0

  val eq_dec : t -> t -> bool
 end
```
(looks good)
and
``module StringOT =
 struct
  type t = char list
…
```
  (** val compare : char list -> char list -> char list compare0 **) 
 
  let rec compare s s2 =
```
Note that the type signature is commented out. The inferred type is going to be more general, because `comare0` has a phantom type parameter, which explains the error message above.

But no idea how to fix this :-)

**Comment by Lysxia** [comment]
I can't reproduce this. OCaml should be able to specialize functions when type-checking modules.

**Comment by nomeata** [comment]
Yeah, I also tried to reproduce it with a small example, by appending this
```
Module Make (A : OrderedType).
End Make.
Require Import Coq.Structures.DecidableTypeEx.
Module Foo := OT_as_DT(StringOT).
Require Extraction.
Extraction "foo.ml" Foo.
```
to `StrintOT.v`, but the thus produced file does *not* have this problem. Guess I need to minimize our problem manually…

**Comment by nomeata** [comment]
Or if someone wants to debug this: Uncommenting the code at the end of https://github.com/antalsz/hs-to-coq/blob/087aa7813874f6f4d8bda832a82c0de4448cda9d/examples/ghc/theories/ContainerAxioms.v shows the problem.

**Comment by Lysxia** [comment]
`char` is shadowed. `QuickChick` uses OCaml's primitive `char`, while `base/GHC/Char.v`  defines the type `Definition Char := N` which extracts to `type char = int`.

**Comment by Lysxia** [comment]
A quick workaround is to declare `Extract Inlined Constant Char => "int".`
